org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java

```diff
@@ -131,26 +161,59 @@ public boolean equals(Object obj) {
             return false;
         }
         DefaultGradleBuild other = (DefaultGradleBuild) obj;
-        return Objects.equals(this.gradleBuild, other.gradleBuild);
+        return Objects.equals(this.buildConfig, other.buildConfig);
     }
 
     /**
      * Executes the synchronization on the target Gradle build.
      */
     private static class SynchronizeOperation extends BaseToolingApiOperation {
 
-        private final org.eclipse.buildship.core.internal.workspace.GradleBuild gradleBuild;
+        private final InternalGradleBuild gradleBuild;
         private final NewProjectHandler newProjectHandler;
+        private List<SynchronizationFailure> failures;
 
-        public SynchronizeOperation(org.eclipse.buildship.core.internal.workspace.GradleBuild gradleBuild, NewProjectHandler newProjectHandler) {
+        public SynchronizeOperation(InternalGradleBuild gradleBuild, NewProjectHandler newProjectHandler) {
             super("Synchronize project " + gradleBuild.getBuildConfig().getRootProjectDirectory().getName());
             this.gradleBuild = gradleBuild;
             this.newProjectHandler = newProjectHandler;
         }
 
+        SynchronizationResult run(CancellationTokenSource tokenSource, IProgressMonitor monitor) {
+            GradleMarkerManager.clear(this.gradleBuild);
+            DefaultSynchronizationResult result;
+            try {
+                CorePlugin.operationManager().run(this, tokenSource, monitor);
+                this.failures.forEach(f -> GradleErrorMarker.create(f.getResource(), this.gradleBuild, f.getMessage(), f.getException(), 0));
+                result = DefaultSynchronizationResult.from(getFailures());
+            } catch (CoreException e) {
+                ToolingApiStatus status = ToolingApiStatus.from("Project synchronization" , e);
+                if (status.severityMatches(IStatus.WARNING | IStatus.ERROR)) {
+                    GradleMarkerManager.addError(this.gradleBuild, status);
```

Comment:
Should make a difference between warning and error. Or maybe the method name is just misleading?

Reply:
No, it actually adds a marker with error severity. I'll add a separate method for defining warning markers.